waiter/src/waiter/reporter.clj

```diff
@@ -0,0 +1,133 @@
+  (:import (com.codahale.metrics ConsoleReporter MetricFilter ScheduledReporter)
+           (com.codahale.metrics.graphite Graphite GraphiteReporter GraphiteSender PickledGraphite)
+           java.io.PrintStream
+           java.net.InetSocketAddress
+           java.util.concurrent.TimeUnit))
+
+(defprotocol CodahaleReporter
+  "A reporter for codahale metrics"
+
+  (close! [this]
+    "Stops the reporter and performs any cleanup")
+
+  (state [this]
+    "Returns the state of this reporter"))
+
+(defn- make-metrics-filter
+  [filter-regex]
+  (reify MetricFilter (matches [_ name _] (some? (re-matches filter-regex name)))))
+
+(defn- make-codahale-reporter
+  [^ScheduledReporter scheduled-reporter state-atom period-ms]
+  (.start scheduled-reporter period-ms TimeUnit/MILLISECONDS)
+  (swap! state-atom merge [:run-state :started])
+  (reify CodahaleReporter
+    (close! [_] (.close scheduled-reporter) (swap! state-atom [:run-state :closed]))
+    (state [_] @state-atom)))
+
+(defn validate-console-reporter-config
+  "Validates ConsoleReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/required-key :period-ms) schema/positive-int
+               (s/optional-key :filter-regex) s/Regex
+               s/Any s/Any} config))
+
+(defn make-console-reporter
+  "Creates a ConsoleReporter for codahale metrics"
+  ([filter-regex] (make-console-reporter filter-regex nil))
+  ([filter-regex ^PrintStream output]
+   (let [state-atom (atom {:run-state :created})
+         console-reporter (-> (cond-> (ConsoleReporter/forRegistry metrics/default-registry)
+                                      output (.outputTo output)
+                                      filter-regex (.filter (make-metrics-filter filter-regex)))
+                              (.convertRatesTo TimeUnit/SECONDS)
+                              (.convertDurationsTo TimeUnit/MILLISECONDS)
+                              (.build))]
+     [console-reporter state-atom])))
+
+(defn console-reporter
+  "Creates and starts a ConsoleReporter for codahale metrics"
+  [config]
+  (let [{:keys [period-ms filter-regex]} (validate-console-reporter-config config)
+        [console-reporter state-atom] (make-console-reporter filter-regex)]
+    (make-codahale-reporter console-reporter state-atom period-ms)))
+
+(defn validate-graphite-reporter-config
+  "Validates GraphiteReporter settings and sets defaults"
+  [config]
+  (s/validate {(s/optional-key :filter-regex) s/Regex
+               (s/required-key :host) s/Str
+               (s/required-key :period-ms) schema/positive-int
+               (s/required-key :pickled?) s/Bool
+               (s/optional-key :prefix) s/Str
+               (s/required-key :port) schema/positive-int
+               s/Any s/Any}
+              (merge {:pickled? true} config)))
+
+(defn make-graphite-reporter
+  "Creates a GraphiteReporter for metrics"
+  ([filter-regex prefix host port pickled?]
+   (let [addr (InetSocketAddress. ^String host ^int port)
```

Comment:
Maybe move the type hints to the parameter list above?

Reply:
CompilerException java.lang.IllegalArgumentException: fns taking primitives support only 4 or fewer args